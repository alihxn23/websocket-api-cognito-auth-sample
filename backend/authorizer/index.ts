// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import * as jwt from "jsonwebtoken";
import axios from "axios";
import jwkToPem from "jwk-to-pem";
import { APIGatewayRequestAuthorizerHandler } from "aws-lambda";

const UserPoolId = process.env.USER_POOL_ID!;
const UserPoolRegion = process.env.USER_POOL_REGION!;
const AppClientId = process.env.APP_CLIENT_ID!;
const jwkCache: { [key: string]: jwkToPem.JWK } = {};

export const handler: APIGatewayRequestAuthorizerHandler = async (event, context) => {
    try {
        const encodedToken = event.queryStringParameters!.idToken!;
        const token = jwt.decode(encodedToken, { complete: true })!;

        if (typeof token == "string") throw Error("invalid jwt.");

        const jwk = await getJwkByKid(token.payload.iss, token.header.kid);
        const pem = jwkToPem(jwk);
        jwt.verify(encodedToken, pem);
        verifyClaims(token.payload, AppClientId, UserPoolId, UserPoolRegion);

        return allowPolicy(event.methodArn, token.payload);
    } catch (error) {
        console.log(error.message);
        return denyAllPolicy();
    }
};

// https://docs.aws.amazon.com/cognito/latest/developerguide/amazon-cognito-user-pools-using-tokens-verifying-a-jwt.html
const verifyClaims = (jwtPayload: any, appClientId: string, userPoolId: string, userPoolRegion: string) => {
    if (jwtPayload.exp * 1000 < Date.now()) {
        throw Error(`token expired.`);
    }

    const iss = jwtPayload.iss;
    if (iss != `https://cognito-idp.${userPoolRegion}.amazonaws.com/${userPoolId}`) {
        throw Error(`wrong iss. ${iss}`);
    }

    const aud = jwtPayload.aud;
    if (aud != appClientId) {
        throw Error(`wrong aud. ${aud}`);
    }

    const token_use = jwtPayload.token_use;
    if (token_use != "id") {
        throw Error(`wrong token_use. ${token_use}`);
    }
};

const getJwkByKid = async (iss: string, kid: string) => {
    if (kid in jwkCache) {
        return jwkCache[kid];
    }

    const jwksEndpoint = iss + "/.well-known/jwks.json";
    const json = await axios.get(jwksEndpoint);
    for (const key of json.data.keys) {
        jwkCache[key.kid] = key;
    }

    return jwkCache[kid];
};

const denyAllPolicy = () => {
    return {
        principalId: "*",
        policyDocument: {
            Version: "2012-10-17",
            Statement: [
                {
                    Action: "*",
                    Effect: "Deny",
                    Resource: "*",
                },
            ],
        },
    };
};

const allowPolicy = (methodArn: string, idToken: any) => {
    return {
        principalId: idToken.sub,
        policyDocument: {
            Version: "2012-10-17",
            Statement: [
                {
                    Action: "execute-api:Invoke",
                    Effect: "Allow",
                    Resource: methodArn,
                },
            ],
        },
        context: {
            // set userId in the context
            userId: idToken.sub,
        },
    };
};
