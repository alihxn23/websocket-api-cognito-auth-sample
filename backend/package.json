{
  "dependencies": {
    "aws-lambda": "^1.0.6",
    "axios": "^0.21.1",
    "jsonwebtoken": "^8.5.1",
    "jwk-to-pem": "^2.0.5",
    "aws-sdk": "^2.856.0",
    "jest": "^26.4.2",
    "ts-jest": "^26.5.3",
    "ts-node": "^9.0.0",
    "typescript": "~3.9.7"
  },
  "devDependencies": {
    "@types/aws-lambda": "^8.10.72",
    "@types/jest": "^26.0.10",
    "@types/jsonwebtoken": "^8.5.0",
    "@types/jwk-to-pem": "^2.0.0",
    "@types/node": "^14.14.37"
  }
}
