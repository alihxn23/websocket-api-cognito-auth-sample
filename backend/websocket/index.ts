// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import * as aws from "aws-sdk";
import { APIGatewayProxyHandler } from "aws-lambda";
import { AWSError } from "aws-sdk";

const docClient = new aws.DynamoDB.DocumentClient();
const ConnectionTableName = process.env.CONNECTION_TABLE_NAME!;

export const handler: APIGatewayProxyHandler = async (event, context) => {
    const routeKey = event.requestContext.routeKey!;
    const connectionId = event.requestContext.connectionId!;

    if (routeKey == "$connect") {
        const userId = event.requestContext.authorizer?.userId;

        try {
            await docClient
                .put({
                    TableName: ConnectionTableName,
                    Item: {
                        userId: userId,
                        connectionId: connectionId,
                        removedAt: Math.ceil(Date.now() / 1000) + 3600 * 3,
                    },
                })
                .promise();
            return { statusCode: 200, body: "Connected." };
        } catch (err) {
            console.error(err);
            return { statusCode: 500, body: "Connection failed." };
        }
    }
    if (routeKey == "$disconnect") {
        try {
            await removeConnectionId(connectionId);
            return { statusCode: 200, body: "Disconnected." };
        } catch (err) {
            console.error(err);
            return { statusCode: 500, body: "Disconnection failed." };
        }
    }

    const managementApi = new aws.ApiGatewayManagementApi({
        endpoint: `${event.requestContext.domainName}/${event.requestContext.stage}`,
    });

    try {
        await managementApi
            .postToConnection({
                ConnectionId: connectionId,
                Data: JSON.stringify({ message: event.body }),
            })
            .promise();
    } catch (e) {
        const error: AWSError = e;
        if (error.statusCode == 410) {
            await removeConnectionId(connectionId);
        } else {
            throw e;
        }
    }

    return { statusCode: 200, body: "Received." };
};

const removeConnectionId = async (connectionId: string) => {
    return await docClient
        .delete({
            TableName: ConnectionTableName,
            Key: {
                connectionId,
            },
        })
        .promise();
};
