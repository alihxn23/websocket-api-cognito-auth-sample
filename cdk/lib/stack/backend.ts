// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import * as cdk from "aws-cdk-lib";
import * as iot from "aws-cdk-lib/aws-iot";
import { Construct } from "constructs";
import { Auth } from "../construct/auth";
import { Storage } from "../construct/storage";
import { Handler } from "../construct/handler";
import { WebSocket } from "../construct/websocket";

export class BackendStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const auth = new Auth(this, `Auth`);
    const storage = new Storage(this, `Storage`);
    const handler = new Handler(this, `Handler`, {
      userPool: auth.userPool,
      userPoolClient: auth.userPoolClient,
      connectionIdTable: storage.connectionIdTable,
      notificationsTable: storage.notificationsTable,
    });

    const websocket = new WebSocket(this, `Websocket`, {
      authHandler: handler.authHandler,
      websocketHandler: handler.websocketHandler,
      connectHandler: handler.connectHandler,
      disconnectHandler: handler.disconnectHandler,
      sendNotificationHandler: handler.sendNotificationHandler,
      getNotificationHandler: handler.getNotificationHandler,
      sendToTopicHandler: handler.websocketToIotCore,
    });

    websocket.api.grantManageConnections(handler.websocketHandler);
    websocket.api.grantManageConnections(handler.sendNotificationHandler);
    websocket.api.grantManageConnections(handler.getNotificationHandler);
    websocket.api.grantManageConnections(handler.notificationStreamHandler);
    websocket.api.grantManageConnections(handler.iotCoreToWebsocket);

    const iotTopicRuleSql = 'SELECT * from "websocket/outgoing"';
    // const iotTopicRule = new iot.CfnTopicRule(
    //   this, "iotRule", {topicRulePayload: {sql: iotTopicRuleSql, actions: []}}
    // )
    const topicRule = new iot.CfnTopicRule(this, "TopicRule", {
      topicRulePayload: {
        sql: iotTopicRuleSql,
        actions: [{ lambda: { functionArn: handler.iotCoreToWebsocket.functionArn } }],
      },
    });

    handler.iotCoreToWebsocket.addPermission("GrantIotRule", {
      principal: new cdk.aws_iam.ServicePrincipal("iot.amazonaws.com"),
      sourceArn: topicRule.attrArn,
    });

    // handler.websocketToIotCore.addToRolePolicy(new cdk.aws_iam.)

    {
      new cdk.CfnOutput(this, `Region`, {
        value: cdk.Stack.of(this).region,
      });

      new cdk.CfnOutput(this, `UserPoolId`, {
        value: auth.userPool.userPoolId,
      });

      new cdk.CfnOutput(this, `UserPoolWebClientId`, {
        value: auth.userPoolClient.userPoolClientId,
      });

      new cdk.CfnOutput(this, `WebSocketEndpoint`, {
        value: websocket.apiEndpoint,
      });
    }
  }
}
