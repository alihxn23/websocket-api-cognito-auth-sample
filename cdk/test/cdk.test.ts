// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import { SynthUtils } from "@aws-cdk/assert";
import * as cdk from "@aws-cdk/core";
import { BackendStack } from "../lib/stack/backend";

test("Snapshot test", () => {
    const app = new cdk.App();
    const stack = new BackendStack(app, "BackendStack");
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
