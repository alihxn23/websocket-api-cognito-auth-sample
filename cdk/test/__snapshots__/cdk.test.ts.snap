// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test 1`] = `
Object {
  "Outputs": Object {
    "Region": Object {
      "Value": Object {
        "Ref": "AWS::Region",
      },
    },
    "UserPoolId": Object {
      "Value": Object {
        "Ref": "authuserPoolE16D54A3",
      },
    },
    "UserPoolWebClientId": Object {
      "Value": Object {
        "Ref": "authuserPoolclient81547F0C",
      },
    },
    "WebSocketEndpoint": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            Object {
              "Fn::GetAtt": Array [
                "websocketapi6BAADC55",
                "ApiEndpoint",
              ],
            },
            "/prod",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters34ded589685dc9a2b989159a4d00c8d6bd780b0c8639bb14f787c38012b931afArtifactHash8C7C64A2": Object {
      "Description": "Artifact hash for asset \\"34ded589685dc9a2b989159a4d00c8d6bd780b0c8639bb14f787c38012b931af\\"",
      "Type": "String",
    },
    "AssetParameters34ded589685dc9a2b989159a4d00c8d6bd780b0c8639bb14f787c38012b931afS3Bucket39A6B72D": Object {
      "Description": "S3 bucket for asset \\"34ded589685dc9a2b989159a4d00c8d6bd780b0c8639bb14f787c38012b931af\\"",
      "Type": "String",
    },
    "AssetParameters34ded589685dc9a2b989159a4d00c8d6bd780b0c8639bb14f787c38012b931afS3VersionKeyF1D9DB0D": Object {
      "Description": "S3 key for asset version \\"34ded589685dc9a2b989159a4d00c8d6bd780b0c8639bb14f787c38012b931af\\"",
      "Type": "String",
    },
    "AssetParameters923f820b1617ed315e8514af425f37c5d8911522ea9f2da6aae7672d834e9251ArtifactHash4024E7E4": Object {
      "Description": "Artifact hash for asset \\"923f820b1617ed315e8514af425f37c5d8911522ea9f2da6aae7672d834e9251\\"",
      "Type": "String",
    },
    "AssetParameters923f820b1617ed315e8514af425f37c5d8911522ea9f2da6aae7672d834e9251S3Bucket95CE8645": Object {
      "Description": "S3 bucket for asset \\"923f820b1617ed315e8514af425f37c5d8911522ea9f2da6aae7672d834e9251\\"",
      "Type": "String",
    },
    "AssetParameters923f820b1617ed315e8514af425f37c5d8911522ea9f2da6aae7672d834e9251S3VersionKeyE1750CEE": Object {
      "Description": "S3 key for asset version \\"923f820b1617ed315e8514af425f37c5d8911522ea9f2da6aae7672d834e9251\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "authuserPoolE16D54A3": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AccountRecoverySetting": Object {
          "RecoveryMechanisms": Array [
            Object {
              "Name": "verified_phone_number",
              "Priority": 1,
            },
            Object {
              "Name": "verified_email",
              "Priority": 2,
            },
          ],
        },
        "AdminCreateUserConfig": Object {
          "AllowAdminCreateUserOnly": false,
        },
        "AutoVerifiedAttributes": Array [
          "email",
        ],
        "EmailVerificationMessage": "The verification code to your new account is {####}",
        "EmailVerificationSubject": "Verify your new account",
        "SmsVerificationMessage": "The verification code to your new account is {####}",
        "VerificationMessageTemplate": Object {
          "DefaultEmailOption": "CONFIRM_WITH_CODE",
          "EmailMessage": "The verification code to your new account is {####}",
          "EmailSubject": "Verify your new account",
          "SmsMessage": "The verification code to your new account is {####}",
        },
      },
      "Type": "AWS::Cognito::UserPool",
      "UpdateReplacePolicy": "Delete",
    },
    "authuserPoolclient81547F0C": Object {
      "Properties": Object {
        "AllowedOAuthFlows": Array [
          "implicit",
          "code",
        ],
        "AllowedOAuthFlowsUserPoolClient": true,
        "AllowedOAuthScopes": Array [
          "profile",
          "phone",
          "email",
          "openid",
          "aws.cognito.signin.user.admin",
        ],
        "CallbackURLs": Array [
          "https://example.com",
        ],
        "ExplicitAuthFlows": Array [
          "ALLOW_USER_PASSWORD_AUTH",
          "ALLOW_USER_SRP_AUTH",
          "ALLOW_REFRESH_TOKEN_AUTH",
        ],
        "SupportedIdentityProviders": Array [
          "COGNITO",
        ],
        "UserPoolId": Object {
          "Ref": "authuserPoolE16D54A3",
        },
      },
      "Type": "AWS::Cognito::UserPoolClient",
    },
    "handler9A1C0C74": Object {
      "DependsOn": Array [
        "handlerServiceRoleDefaultPolicy6BFD97B0",
        "handlerServiceRole2B8AB3DF",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters34ded589685dc9a2b989159a4d00c8d6bd780b0c8639bb14f787c38012b931afS3Bucket39A6B72D",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters34ded589685dc9a2b989159a4d00c8d6bd780b0c8639bb14f787c38012b931afS3VersionKeyF1D9DB0D",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters34ded589685dc9a2b989159a4d00c8d6bd780b0c8639bb14f787c38012b931afS3VersionKeyF1D9DB0D",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "CONNECTION_TABLE_NAME": Object {
              "Ref": "storageconnectionIdTable7B549566",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "handlerServiceRole2B8AB3DF",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "handlerServiceRole2B8AB3DF": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "handlerServiceRoleDefaultPolicy6BFD97B0": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "storageconnectionIdTable7B549566",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "storageconnectionIdTable7B549566",
                          "Arn",
                        ],
                      },
                      "/index/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "execute-api:ManageConnections",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "websocketapi6BAADC55",
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "handlerServiceRoleDefaultPolicy6BFD97B0",
        "Roles": Array [
          Object {
            "Ref": "handlerServiceRole2B8AB3DF",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "handlerauthorizer4ED01BFC": Object {
      "DependsOn": Array [
        "handlerauthorizerServiceRole02D804DF",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters923f820b1617ed315e8514af425f37c5d8911522ea9f2da6aae7672d834e9251S3Bucket95CE8645",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters923f820b1617ed315e8514af425f37c5d8911522ea9f2da6aae7672d834e9251S3VersionKeyE1750CEE",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters923f820b1617ed315e8514af425f37c5d8911522ea9f2da6aae7672d834e9251S3VersionKeyE1750CEE",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "APP_CLIENT_ID": Object {
              "Ref": "authuserPoolclient81547F0C",
            },
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "USER_POOL_ID": Object {
              "Ref": "authuserPoolE16D54A3",
            },
            "USER_POOL_REGION": Object {
              "Ref": "AWS::Region",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "handlerauthorizerServiceRole02D804DF",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "handlerauthorizerBackendStackwebsocketA340BC86Permissions6A1233C9": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "handlerauthorizer4ED01BFC",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "websocketapi6BAADC55",
              },
              "/authorizers/",
              Object {
                "Ref": "websocketauthorizerA68CA492",
              },
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "handlerauthorizerServiceRole02D804DF": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "storageconnectionIdTable7B549566": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "connectionId",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "userId",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "GlobalSecondaryIndexes": Array [
          Object {
            "IndexName": "userIdIndex",
            "KeySchema": Array [
              Object {
                "AttributeName": "userId",
                "KeyType": "HASH",
              },
            ],
            "Projection": Object {
              "ProjectionType": "ALL",
            },
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "connectionId",
            "KeyType": "HASH",
          },
        ],
        "TimeToLiveSpecification": Object {
          "AttributeName": "removedAt",
          "Enabled": true,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "websocketapi6BAADC55": Object {
      "Properties": Object {
        "Name": "api",
        "ProtocolType": "WEBSOCKET",
        "RouteSelectionExpression": "$request.body.action",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "websocketapiconnectRoute50EA91CF": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "websocketapi6BAADC55",
        },
        "AuthorizationType": "CUSTOM",
        "AuthorizerId": Object {
          "Ref": "websocketauthorizerA68CA492",
        },
        "RouteKey": "$connect",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "websocketapiconnectRouteWebSocketIntegrationa76404738367620e6bbc9be82111ea649E3AB3B8",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "websocketapiconnectRouteBackendStackwebsocketapiconnectRoute0639A138PermissionDA2FF12D": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "handler9A1C0C74",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "websocketapi6BAADC55",
              },
              "/*/*$connect",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "websocketapiconnectRouteWebSocketIntegrationa76404738367620e6bbc9be82111ea649E3AB3B8": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "websocketapi6BAADC55",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": Object {
          "Fn::GetAtt": Array [
            "handler9A1C0C74",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "websocketapidefaultRouteBackendStackwebsocketapidefaultRouteE75433E8Permission167FAA0D": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "handler9A1C0C74",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "websocketapi6BAADC55",
              },
              "/*/*$default",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "websocketapidefaultRouteEDB130FF": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "websocketapi6BAADC55",
        },
        "RouteKey": "$default",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "websocketapiconnectRouteWebSocketIntegrationa76404738367620e6bbc9be82111ea649E3AB3B8",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "websocketapidisconnectRoute1CE296B1": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "websocketapi6BAADC55",
        },
        "RouteKey": "$disconnect",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "websocketapiconnectRouteWebSocketIntegrationa76404738367620e6bbc9be82111ea649E3AB3B8",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "websocketapidisconnectRouteBackendStackwebsocketapidisconnectRouteA44CF8DDPermission103044D4": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "handler9A1C0C74",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "websocketapi6BAADC55",
              },
              "/*/*$disconnect",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "websocketauthorizerA68CA492": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "websocketapi6BAADC55",
        },
        "AuthorizerType": "REQUEST",
        "AuthorizerUri": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":apigateway:",
              Object {
                "Ref": "AWS::Region",
              },
              ":lambda:path/2015-03-31/functions/",
              Object {
                "Fn::GetAtt": Array [
                  "handlerauthorizer4ED01BFC",
                  "Arn",
                ],
              },
              "/invocations",
            ],
          ],
        },
        "IdentitySource": Array [
          "route.request.querystring.idToken",
        ],
        "Name": "lambdaAuthorizer",
      },
      "Type": "AWS::ApiGatewayV2::Authorizer",
    },
    "websocketstage112FEEBD": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "websocketapi6BAADC55",
        },
        "AutoDeploy": true,
        "StageName": "prod",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
  },
}
`;
